src/NuGet.Jobs.RegistrationComparer/HiveComparer.cs

```diff
@@ -0,0 +1,244 @@
+                    rawIndex.Data.ToObject<RegistrationIndex>(NuGetJsonSerialization.Serializer)));
+            }
+
+            // Download the pages (if any) and leaves.
+            var pageUrlGroups = indexes
+                .Select((x, i) => x
+                    .Data
+                    .Items
+                    .Where(p => p.Items == null)
+                    .Select(p => p.Url)
+                    .ToList())
+                .ToList();
+            var leafUrlGroups = baseUrls
+                .Select(x => versions.Select(v => $"{x}{id}/{v}.json").ToList())
+                .ToList();
+
+            var urls = new ConcurrentBag<string>(pageUrlGroups
+                .SelectMany(x => x)
+                .Concat(leafUrlGroups.SelectMany(x => x)));
+            var urlToJson = new ConcurrentDictionary<string, JObject>();
+            await ParallelAsync.Repeat(
+                async () =>
+                {
+                    await Task.Yield();
+                    while (urls.TryTake(out var pageUrl))
+                    {
+                        var json = await GetJObjectOrNullAsync(pageUrl);
+                        urlToJson.TryAdd(pageUrl, json.Data);
+                    }
+                });
+
+            // Compare the pages.
+            for (var i = 1; i < baseUrls.Count; i++)
+            {
+                for (var pageIndex = 0; pageIndex < pageUrlGroups[i].Count; pageIndex++)
+                {
+                    var leftUrl = pageUrlGroups[i - 1][pageIndex];
+                    var rightUrl = pageUrlGroups[i][pageIndex];
+
+                    var comparisonContext = new ComparisonContext(
+                        id,
+                        baseUrls[i - 1],
+                        baseUrls[i],
+                        leftUrl,
+                        rightUrl,
+                        Normalizers.Page);
+
+                    _comparer.Compare(
+                       urlToJson[leftUrl],
+                       urlToJson[rightUrl],
+                       comparisonContext);
+                }
+            }
+
+            // Compare the affected leaves.
+            for (var i = 1; i < baseUrls.Count; i++)
+            {
+                for (var leafIndex = 0; leafIndex < leafUrlGroups[i].Count; leafIndex++)
+                {
+                    var leftUrl = leafUrlGroups[i - 1][leafIndex];
+                    var rightUrl = leafUrlGroups[i][leafIndex];
+
+                    try
+                    {
+                        if (AreBothMissing(
+                            leftUrl,
+                            rightUrl,
+                            urlToJson[leftUrl],
+                            urlToJson[rightUrl]))
+                        {
+                            continue;
+                        }
+                    }
+                    catch (InvalidOperationException ex)
+                    {
+                        ResultWriter.WriteWarning(ex.Message);
```

Comment:
Nit: consider recording `.ToString()` instead

Reply:
Chose not to in this case since the exception this case is a well-known throw-er. I can do the whole string if you want, though.

Reply:
Its' fine as is :)